packages/server/src/auth/types/request.ts

```diff
@@ -11,15 +12,101 @@ export function isFetchRequest(request: GenericRequest): request is Request {
     return "headers" in request && typeof request.headers.get === "function";
 }
 
-export function isNodeResponse(response: GenericResponse): response is ServerResponse {
-    return "setHeader" in response;
+interface RequestAdapter {
+    getCookieHeader: () => string | null | undefined;
+    getBody: () => Promise<any>;
 }
 
-export function isFetchResponse(response: GenericResponse): response is Response {
-    return "headers" in response && typeof response.headers.append === "function";
+export class NodeRequestAdapter implements RequestAdapter {
+    constructor(private request: IncomingMessage) {}
+
+    getCookieHeader() {
+        return this.request.headers.cookie;
+    }
+
+    getBody() {
+        return getNodeRequestBody(this.request);
+    }
+}
+
+export class FetchRequestAdapter implements RequestAdapter {
+    constructor(private request: Request) {}
+
+    getCookieHeader() {
+        return this.request.headers.get("Cookie");
+    }
+
+    getBody() {
+        return this.request.json();
+    }
+}
+
+export function isNodeResponse(response?: GenericResponse): response is ServerResponse {
+    return response != null && "setHeader" in response;
+}
+
+export function isFetchResponse(response?: GenericResponse): response is Response {
+    return response != null && "headers" in response && typeof response.headers.append === "function";
+}
+
+interface ResponseAdapter {
+    setCookies: (cookies: string[]) => void;
+    setAuthMaterial: (authMaterial: AuthMaterialWithUser) => GenericResponse | null;
+    setError: (statusCode: number, errorResponseBody: ErrorBody) => GenericResponse;
+}
+
+export class NodeResponseAdapter implements ResponseAdapter {
+    constructor(private response?: ServerResponse) {}
+
+    setCookies(cookies: string[]) {
+        if (this.response == null) {
+            throw new Error("Response not found");
+        }
+        this.response.setHeader("Set-Cookie", cookies);
```

Comment:
So Node supports sending an array of cookies here and it will handle setting multiple `Set-Cookie` headers, while fetch needs each cookie to be set separately, right?